### ISSUE TITLE
cd docs 2

### ISSUE THREAD

**Issue by SkBlaz** [SkBlaz/py3plex#820]
You are GitHub Copilot helping me REWRITE the py3plex documentation page:

  "How to Run Community Detection on Multilayer Networks"
  (the file that currently renders to https://skblaz.github.io/py3plex/how-to/run_community_detection.html).

GOAL
=====

Upgrade this page from “basic examples” to a *serious, research-grade* how-to that:

1. Presents community detection as a **core analysis workflow** in py3plex, not just a few function calls.
2. Explains **what the algorithms are actually doing** (Louvain, multilayer Louvain, Infomap, label propagation) with light formalism.
3. Makes the **multilayer aspect explicit**: what “communities in multilayer networks” actually mean and how coupling works.
4. Shows **real outputs** for the main workflows (number of communities, size distributions, modularity values, NMI/ARI between algorithms, etc.).
5. Documents how to **analyze and export** community assignments in a robust way (DataFrames, CSVs, visualizations).
6. Stays 100% consistent with the **real py3plex API** (inspect the code, don’t invent functions or modules).

CONTEXT & FILE
===============

- You are editing the existing how-to page for community detection, currently structured like this:

  - Goal & prerequisites
  - Quick Start: Louvain
  - Multilayer-Specific: Multilayer Louvain
  - Infomap Algorithm
  - Label Propagation
  - Analyzing Community Structure (count sizes, visualize, export)
  - Compare Algorithms
  - Layer-Specific Communities
  - Cross-Layer Community Analysis
  - Quality Metrics (modularity)
  - Next Steps

- The current examples mostly just **print number of communities** and a few lines. That feels toy-like.
- We want to preserve the *menu structure and anchors* (headings and slugs), but make the content under each heading richer, more formal, and with realistic outputs.

- IMPORTANT: Use the *actual* py3plex API as in the repo and docs:

  - e.g. `from py3plex.algorithms.community_detection import louvain_communities, infomap_communities, label_propagation, calculate_modularity`
  - `from py3plex.algorithms.community_detection.multilayer_louvain import multilayer_louvain` (or whatever the real entry point/module/file name is).
  - `from py3plex.core import multinet`
  - `from py3plex.dsl import Q, L` etc.

  Verify imports and function names against the codebase — don’t guess.

HIGH-LEVEL STRUCTURE
=====================

Keep the same top-level sections, but enrich each:

1. **Goal & Prerequisites** (already present; refine wording)
2. **Quick Start: Louvain Algorithm**
3. **Multilayer-Specific: Multilayer Louvain**
4. **Infomap Algorithm**
5. **Label Propagation**
6. **Analyzing Community Structure**
   - Count Nodes Per Community
   - Visualize Communities
   - Export Communities
7. **Compare Algorithms**
8. **Layer-Specific Communities**
9. **Cross-Layer Community Analysis**
10. **Quality Metrics**
    - Compute Modularity
11. **(Optional) CLI Cross-Reference**
12. **Next Steps**

For each section, make the examples look like *real workflows* and embed small “Expected output” blocks.

DETAILED GUIDELINES PER SECTION
================================

### 1. Goal & Prerequisites

Keep the “Goal” and “Prerequisites”, but:

- Expand “Goal” into 2–3 sentences explaining what community detection means in the context of *multilayer networks*:

  - Communities = groups of nodes that are more densely connected within the group than to the rest.
  - In multilayer networks, communities can:

    - Live within a single layer.
    - Span multiple layers.
    - Be influenced by inter-layer coupling (e.g., node identity across layers).

- Add one short paragraph on **when and why** to use community detection:
  - Understanding mesoscale structure, detecting functional modules, etc.

### 2. Quick Start: Louvain Algorithm

The “Quick Start” example should feel like something you’d actually run in a paper or analysis:

- Keep the basic code structure:

  ```python
  from py3plex.core import multinet
  from py3plex.algorithms.community_detection import louvain_communities

  network = multinet.multi_layer_network()
  network.load_network("data.multiedgelist", input_type="multiedgelist")

  communities = louvain_communities(network)

Add a short explanation of what Louvain does:

Maximizes modularity on the aggregated graph (or on an internal projection).

Non-overlapping hard partition.

Fast and usually a good default for large networks.


Show more informative output:

Keep the print(f"Found {len(set(communities.values()))} communities").

Add a snippet computing and printing:

Number of communities.

Top 5 community sizes (using collections.Counter).

Example nodes in the largest community.



Example pattern:

from collections import Counter

comm_ids = list(communities.values())
num_comms = len(set(comm_ids))
print(f"Found {num_comms} communities")

sizes = Counter(comm_ids)
print("Top communities by size:")
for comm_id, size in sizes.most_common(5):
    print(f"  Community {comm_id}: {size} nodes")

Embed an Expected output block that looks realistic:

Something like:

Found 4 communities
Top communities by size:
  Community 0: 42 nodes
  Community 1: 27 nodes
  Community 2: 19 nodes
  Community 3: 12 nodes


Make clear in prose that exact numbers depend on the dataset.


3. Multilayer-Specific: Multilayer Louvain

This section is the heart of multilayer community detection — expand it significantly:

Explain what multilayer modularity is in words:

Communities are detected jointly across layers.

There is an inter-layer coupling parameter (often called ω) that ties node replicas across layers.

Higher ω → stronger preference for nodes to be in the same community across layers.


Briefly reference the idea of modularity:

For single layer, modularity (Q) measures how much more dense the edges inside communities are than expected under a null model.

For multilayer, modularity sums contributions from each layer plus coupling between node copies across layers.


Keep omega (or whatever is actually called) and explain its role:

omega=0: each layer independent.

omega > 0: cross-layer coupling; communities tend to align across layers.

Show at least two values (e.g. 0, 0.5, 1.0) in a small code snippet to illustrate the effect.


Expand the code example:

from py3plex.algorithms.community_detection.multilayer_louvain import multilayer_louvain
from py3plex.algorithms.community_detection import calculate_modularity
from collections import Counter

for omega in [0.0, 0.5, 1.0]:
    communities = multilayer_louvain(
        network,
        omega=omega  # Inter-layer coupling
    )

    num_comms = len(set(communities.values()))
    sizes = Counter(communities.values())
    modularity = calculate_modularity(network, communities)

    print(f"omega={omega:.1f} → {num_comms} communities, modularity={modularity:.3f}")
    print("  Largest community size:", sizes.most_common(1)[0][1])

Embed an Expected output block such as:

omega=0.0 → 7 communities, modularity=0.412
  Largest community size: 18
omega=0.5 → 5 communities, modularity=0.529
  Largest community size: 26
omega=1.0 → 4 communities, modularity=0.547
  Largest community size: 33

(Use realistic but not necessarily exact values.)

Add 2–3 sentences interpreting this: increasing omega reduces the number of communities but increases cross-layer coherence and often modularity.


4. Infomap Algorithm

Make Infomap feel less like a black box:

Briefly describe Infomap:

Flow-based / information-theoretic method.

Minimizes the description length of a random walk on the network.

Sensitive to edge weights and directions (if supported in the implementation).


Expand the code example:

from py3plex.algorithms.community_detection import infomap_communities
from collections import Counter

communities = infomap_communities(network)

num_comms = len(set(communities.values()))
print(f"Found {num_comms} communities with Infomap")

sizes = Counter(communities.values())
print("Top communities (Infomap):")
for comm_id, size in sizes.most_common(5):
    print(f"  Community {comm_id}: {size} nodes")

Show a small Expected output block, and optionally compare to Louvain:

Found 6 communities with Infomap
Top communities (Infomap):
  Community 0: 30 nodes
  Community 1: 22 nodes
  ...

Add a short note: Infomap sometimes finds more/finer communities than Louvain, especially in networks with directional or flow-like structure.


5. Label Propagation

Similarly, expand this section beyond a single print:

Describe label propagation:

Very fast, especially for large networks.

Iteratively updates node labels based on majority labels of neighbors.

May be unstable / non-deterministic depending on update order and random seed.


Update the snippet:

from py3plex.algorithms.community_detection import label_propagation
from collections import Counter

communities = label_propagation(network)

num_comms = len(set(communities.values()))
print(f"Found {num_comms} communities with label propagation")

sizes = Counter(communities.values())
print("Top communities (Label Propagation):")
for comm_id, size in sizes.most_common(5):
    print(f"  Community {comm_id}: {size} nodes")

Optionally mention reproducibility: if label propagation uses randomness, suggest re-running with different seeds and measuring stability.


6. Analyzing Community Structure

Upgrade this from a few lines to a small “analysis toolbox”:

Count Nodes Per Community

Keep using Counter(communities.values()).

Show:

Sorting by size, printing top-k communities.

Optionally compute basic statistics: mean, median community size, etc.


Add an Expected output snippet showing a list of community sizes.


Visualize Communities

Ensure visualize_communities is a real function. If not, adjust imports to the actual visualization API.

Show a more detailed call:

from py3plex.visualization import visualize_communities

visualize_communities(
    network,
    communities,
    output_file="communities.png",
    show=True,
    layout="force",  # only if supported; otherwise omit
)

Add a short prose description of what the figure shows: nodes colored by community, layers separated, etc.

If Sphinx supports embedding the generated image in the docs, wire it similarly to other plots.


Export Communities

Keep the DataFrame pattern, but:

Clarify what keys look like: (node, layer) vs node ID only.

Optionally add the option to export also community sizes.


Example:

import pandas as pd

data = []
for (node, layer), comm_id in communities.items():
    data.append({
        "node": node,
        "layer": layer,
        "community": comm_id,
    })

df = pd.DataFrame(data)
print(df.head())
df.to_csv("communities.csv", index=False)

Add a short text “Expected output (first few rows)” showing a 4–5 row table-like block.


7. Compare Algorithms

This section is ideal for showing that the page is not a toy.

Keep the Adjusted Rand Index example, but expand:

Run at least Louvain and Infomap (and optionally label propagation) and compute pairwise adjusted Rand indices.


from py3plex.algorithms.community_detection import (
    louvain_communities,
    infomap_communities,
    label_propagation,
)
from sklearn.metrics import adjusted_rand_score

louvain_comms = louvain_communities(network)
infomap_comms = infomap_communities(network)
label_prop_comms = label_propagation(network)

nodes = list(louvain_comms.keys())

def labels_from_dict(comm_dict, nodes):
    return [comm_dict[n] for n in nodes]

louvain_labels = labels_from_dict(louvain_comms, nodes)
infomap_labels = labels_from_dict(infomap_comms, nodes)
label_prop_labels = labels_from_dict(label_prop_comms, nodes)

ari_louvain_infomap = adjusted_rand_score(louvain_labels, infomap_labels)
ari_louvain_labelprop = adjusted_rand_score(louvain_labels, label_prop_labels)

print(f"Louvain vs Infomap ARI: {ari_louvain_infomap:.3f}")
print(f"Louvain vs Label Propagation ARI: {ari_louvain_labelprop:.3f}")

Add an Expected output block with plausible ARI values:

Louvain vs Infomap ARI: 0.732
Louvain vs Label Propagation ARI: 0.615

Briefly interpret: ARI near 1 means strong agreement; near 0 means random-like agreement.


8. Layer-Specific Communities

This section is already good conceptually; polish and deepen it:

Provide a sentence explaining why you might want layer-specific communities:

E.g., discovering role-based clusters in the "work" layer vs “friends” layer.


Clarify what Q and L from the DSL are doing:

Q.edges().from_layers(L[layer]).execute(network) extracts a single-layer subgraph.


Keep the code pattern, but ensure:

The object returned by the DSL is compatible with louvain_communities.

If not, show how to convert/unwrap as needed.


Add an Expected output snippet:

Found 3 communities in friends

Optionally show how to compare layer-specific communities to multilayer communities (e.g., using ARI restricted to that layer).


9. Cross-Layer Community Analysis

This is a powerful section — make it less cryptic:

Explain conceptually:

You run community detection per layer and then measure how stable communities are across layers using NMI or ARI.

This helps answer: “Do the same groups of nodes stick together across layers?”


Ensure that the code uses a valid NMI function from scikit-learn:

e.g., from sklearn.metrics import normalized_mutual_info_score as nmi.


Keep the nested loop over pairs of layers, but:

Add a short explanation before the code that:

You first compute layer-wise partitions.

Then for each layer pair, you restrict attention to nodes present in both.



Add an example Expected output block:

friends vs work: NMI = 0.684
friends vs social: NMI = 0.512
work vs social: NMI = 0.431

Add a sentence interpreting these values: high NMI ⇒ communities are similar across layers; low NMI ⇒ layer-specific structure.


10. Quality Metrics (Compute Modularity)

Elevate this beyond a single print:

Describe modularity briefly:

Values roughly in [-0.5, 1].

Higher values (e.g., >0.3) often indicate meaningful community structure, but interpretation is dataset-dependent.


Show:

from py3plex.algorithms.community_detection import calculate_modularity

modularity = calculate_modularity(network, communities)
print(f"Modularity: {modularity:.3f}")

Add Expected output:

Modularity: 0.542

Optionally, show how to compare modularity across algorithms:

for name, comms in [
    ("Louvain", louvain_comms),
    ("Infomap", infomap_comms),
    ("Label Propagation", label_prop_comms),
]:
    q = calculate_modularity(network, comms)
    print(f"{name} modularity: {q:.3f}")

With a small Expected output listing modularity values by method.


11. (Optional) CLI Cross-Reference

Add a small callout (short subsection or note) that:

Mentions the CLI community detection command:

py3plex community net.graphml --algorithm louvain --output communities.json


Clarifies that the how-to page focuses on Python API, but the CLI offers similar functionality for non-programmatic workflows.

Link to the CLI documentation page.


12. Next Steps

Leave the existing bullet list, but:

Add one sentence for each link that explains why you’d click it:

Visualization → “Learn how to visualize node colors by community across layers.”

Algorithm Landscape → “Understand when to choose Louvain vs Infomap vs others.”

Algorithm Roadmap → “See the full list of community detection functions and parameters.”



STYLE & TONE

Technical, concise, but not dry. Avoid “toy tutorial” tone.

Use short paragraphs and bullet points; avoid walls of text.

Where possible, include Expected output blocks that look like real console output or DataFrame heads.

Use consistent terminology:

“community assignments” vs “partition” vs “clusters”.

“multilayer” vs “multiplex” — follow the project’s preferred wording for this page.



IMPLEMENTATION NOTES

Do not invent new public functions, arguments, or modules.

Only use functions that exist in py3plex.algorithms.community_detection, py3plex.algorithms.community_detection.multilayer_louvain, py3plex.visualization, and related modules.

If the current doc uses an import that is wrong or outdated, fix it to the actual API.


For each code block:

Make sure it is runnable given a standard py3plex installation and typical example data.

When you show “Expected output”, use realistic numbers and structure, but it doesn’t have to match exactly.



Now, go through the existing run_community_detection how-to file and rewrite/expand it according to all of the above guidelines, keeping the existing top-level headings but significantly enriching the explanations, code snippets, and embedded outputs so the page looks like a serious, research-ready reference instead of a toy example.
